docs/Testing.md

```diff
@@ -136,79 +46,103 @@ try {
-if (require.main === module) {
-  // Run the server with:
-  // $ node server.js -p 8080
-  start(minimist(process.argv.slice(2), {
-    integer: ['port'],
-    alias: {
-      port: 'p'
-    },
-    default: {
-      port: 3000
-    }
-  }), (err, instance) => {
-    if (err) throw err
-
-    console.log(`server listening on ${instance.server.address().port}`)
+function buildFastify () {
+  const fastify = Fastify()
+
+  fastify.get('/', function (request, reply) {
+    reply.send({ hello: 'world' })
   })
+  
+  return fastify
 }
 
-// note that now we are also exposing the fastify instance
-module.exports = { start, fastify }
+module.exports = { buildFastify }
 ```
 
+**test.js**
 ```js
-// test.js
-const t = require('tap')
-const test = t.test
-const fastify = require('./server').fastify
+const tap = require('tap')
+const buildFastify = require('./app').buildFastify
 
-test('GET `/` route', t => {
-  t.plan(3)
+tap.test('GET `/` route', t => {
+  t.plan(4)
+  
+  const fastify = buildFastify()
 
   fastify.inject({
     method: 'GET',
     url: '/'
-  }, (err, res) => {
+  }, (err, response) => {
     t.error(err)
-    t.strictEqual(res.statusCode, 200)
-    t.strictEqual(res.headers['content-length'], '' + res.payload.length)
-    t.deepEqual(JSON.parse(res.payload), { hello: 'world' })
-    // even if the server is not running (inject does not run the server)
-    // at the end of your tests is highly recommended call `.close()`,
-    // in this way you will close all the connections to external services
+    t.strictEqual(response.statusCode, 200)
+    t.strictEqual(response.headers['content-type'], 'application/json')
+    t.deepEqual(JSON.parse(response.payload), { hello: 'world' })
+    // Even if the server is not running (inject does not run the server),
+    // at the end of your tests it is highly recommended to call `.close()`
+    // to ensure that all connections to external services get closed.
     fastify.close()
   })
 })
 ```
+
+### Testing with a running server
+Fastify can also be tested after starting the server with `fastify.listen()` or after initializing routes and plugins with `fastify.ready()`.
+
+#### Example:
+
+Uses **app.js** from the previous example.
+
+**test-listen.js** (testing with [`Request`](https://www.npmjs.com/package/request))
+```js
+const tap = require('tap')
+const request = require('request')
+const buildFastify = require('./app').buildFastify
+
+tap.test('GET `/` route', t => {
+  t.plan(5)
+  
+  const fastify = buildFastify()
+  
+  fastify.listen(0, (err) => {
+    t.error(err)
+    
+    request({
+      method: 'GET',
+      url: 'http://localhost:' + fastify.server.address().port
+    }, (err, response, body) => {
+      t.error(err)
+      t.strictEqual(response.statusCode, 200)
+      t.strictEqual(response.headers['content-type'], 'application/json')
+      t.deepEqual(JSON.parse(body), { hello: 'world' })
+      fastify.close()
```

Comment:
we should be using t.teardown(fastify.close.bind(fastify))

Reply:
Even though it's slower, I used `t.tearDown(() => fastify.close())` because it's shorter and more readable.